src/librustc_typeck/check/method/probe.rs

```diff
@@ -227,7 +261,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             })
             .collect();
 
-        let final_ty = autoderef.unambiguous_final_ty();
+        let final_ty = match looking_for {
+            &LookingFor::MethodName(_) => autoderef.unambiguous_final_ty(),
+            &LookingFor::ReturnType(_) => self_ty,
+        };
```

Comment:
More operations are performed in the case of `ReturnType` on the returned values and we need all information on the type, which isn't the case for `MethodName`.